core/scheduler/src/main/scala/org/apache/openwhisk/core/scheduler/container/ContainerManager.scala

```diff
@@ -462,32 +549,29 @@ object ContainerManager {
     list
   }
 
-  private def chooseInvokerFromCandidates(
-    candidates: List[InvokerHealth],
-    wholeInvokers: List[InvokerHealth],
-    pairs: List[ScheduledPair],
-    msg: ContainerCreationMessage)(implicit logging: Logging): (List[ScheduledPair], List[InvokerHealth]) = {
-    val idx = rng(mod = candidates.size)
-    val instance = candidates(idx)
-    // it must be compared to the instance unique id
-    val idxInWhole = wholeInvokers.indexOf(wholeInvokers.filter(p => p.id.instance == instance.id.instance).head)
-    val requiredMemory = msg.whiskActionMetaData.limits.memory.megabytes
-    val updated =
-      if (instance.id.userMemory.toMB - requiredMemory >= requiredMemory) { // Since ByteSize is negative, it converts to long type and compares.
-        wholeInvokers.updated(
-          idxInWhole,
-          instance.copy(id = instance.id.copy(userMemory = instance.id.userMemory - requiredMemory.MB)))
+  protected[container] def chooseInvokerFromCandidates(candidates: List[InvokerHealth], msg: ContainerCreationMessage)(
+    implicit logging: Logging): ScheduledPair = {
+    val requiredMemory = msg.whiskActionMetaData.limits.memory
+    if (candidates.isEmpty) {
+      ScheduledPair(msg, invokerId = None, Some(NoAvailableInvokersError), Some(s"No available invokers."))
+    } else if (candidates.forall(p => p.id.userMemory.toMB < requiredMemory.megabytes)) {
+      ScheduledPair(msg, invokerId = None, Some(NoAvailableInvokersError), Some(s"No available invokers."))
+    } else {
+      val idx = rng(mod = candidates.size)
+      val instance = candidates(idx)
+      if (instance.id.userMemory.toMB < requiredMemory.megabytes) {
+        val split = candidates.splitAt(idx)
+        val _ :: t1 = split._2
+        chooseInvokerFromCandidates(split._1 ::: t1, msg)
       } else {
-        // drop the nth element
-        val split = wholeInvokers.splitAt(idxInWhole)
```

Comment:
It was also a bug.
When the chosen invoker does not have enough memory, it just removed the invoker from the list but still, the message is scheduled to the invoker without enough resources.